caffe2/operators/top_k.cu

```diff
@@ -16,216 +16,399 @@
+    const int k,
+    T* values,
+    TIndex* indices,
+    CUDAContext* context) {
+  constexpr int kBlockSize = 256;
+  constexpr int kNumWarps = kBlockSize / kWarpSize;
+  constexpr int smem = kNumWarps * kHeapSize * (sizeof(T) + sizeof(TIndex));
+  constexpr T kInitVal = kSelectMax ? std::numeric_limits<T>::lowest()
+                                    : std::numeric_limits<T>::max();
+  selectRowsViaHeap<T, TIndex, TIndex, kBlockSize, kHeapSize, kSelectMax>
+      <<<outer_size, kBlockSize, smem, context->cuda_stream()>>>(
+          input,
+          values,
+          indices,
+          kInitVal,
+          std::numeric_limits<TIndex>::max(),
+          outer_size,
+          inner_size,
+          k);
+}
+
+template <typename T, bool kSelectMax = true>
+void RunRadixSelectionImpl(
+    const T* input,
+    const TIndex outer_size,
+    const TIndex inner_size,
+    const int k,
+    T* values,
+    TIndex* indices,
+    CUDAContext* context) {
+  const int block = std::min(
+      math::roundUp(static_cast<int>(inner_size), kWarpSize),
+      CAFFE_CUDA_NUM_THREADS);
+  gatherTopK<T, kSelectMax, TIndex>
+      <<<outer_size, block, 0, context->cuda_stream()>>>(
+          input, inner_size, k, outer_size, values, indices);
+  // Unfortunately the output is not currently sorted, and there is no batch
+  // sorting utility available. Iterate over all of the slices and sort them
+  // in-place using Thrust.
+  for (int i = 0; i < outer_size; ++i) {
+    thrust::sort_by_key(
+        thrust::cuda::par.on(context->cuda_stream()),
+        values + i * k,
+        values + i * k + k,
+        indices + i * k,
+        thrust::greater<T>());
+  }
+}
+
+template <typename T>
+void RunTopKOnLastDimCUDAImpl(
+    const T* input,
+    const TIndex outer_size,
+    const TIndex inner_size,
+    const int k,
+    T* values,
+    TIndex* indices,
+    CUDAContext* context) {
+  if (k <= 512) {
+    // k is small, uses heap selection.
+    if (k <= 32) {
+      RunHeapSelectionImpl<T, 32>(
+          input, outer_size, inner_size, k, values, indices, context);
+    } else if (k <= 128) {
+      RunHeapSelectionImpl<T, 128>(
+          input, outer_size, inner_size, k, values, indices, context);
+    } else {
+      RunHeapSelectionImpl<T, 512>(
+          input, outer_size, inner_size, k, values, indices, context);
     }
+  } else {
+    RunRadixSelectionImpl<T>(
+        input, outer_size, inner_size, k, values, indices, context);
+  }
+}
+
+__global__ void FlattenIndicesCUDA(
+    const TIndex* src,
+    const TIndex size,
+    const TIndex stride,
+    const TIndex n,
+    const int k,
+    TIndex* dst) {
+  CUDA_1D_KERNEL_LOOP(i, size) {
+    const TIndex x = i / stride / k;
+    const TIndex y = i % stride;
+#if __CUDA_ARCH__ >= 350
+    dst[i] = __ldg(src + i) * stride + x * n * stride + y;
+#else
+    dst[i] = src[i] * stride + x * n * stride + y;
+#endif
+  }
+}
+
+template <typename T>
+__global__ void SetTopKGradientCUDA(
```

Comment:
Nit: briefly add some simple comment about the core algorithm/formula. It will be very helpful.